### ISSUE TITLE
targetPort does not accept integer as valid

### ISSUE THREAD

**Issue by brandonkal** [Azure/vscode-kubernetes-tools#632]
targetPort can be an integer, but VSCode reports a linting error that it expected a "string."

```yaml
apiVersion: v1
kind: Service
metadata:
  name: hello
spec:
  selector:
    app: hello
  type: ClusterIP
  ports:
    - protocol: TCP
      port: 80
      targetPort: 8080
```

**Comment by itowlson** [comment]
This seems to be an issue with what Kubernetes reports to us as the OpenAPI specification for the cluster.  I'll take a look; we might need to do some massaging.  Thanks for reporting it!

**Comment by itowlson** [comment]
Related: #618, #621,

**Comment by itowlson** [comment]
I think the problem is this in the k8s Swagger `definitions` section:

```
    "io.k8s.apimachinery.pkg.util.intstr.IntOrString": {
      "description": "IntOrString is a type that can hold an int32 or a string.  When used in JSON or YAML marshalling and unmarshalling, it produces or consumes the inner type.  This allows you to have, for example, a JSON field that can accept a name or number.",
      "type": "string",
      "format": "int-or-string"
    },
```

The Red Hat YAML handler is seeing the `type: string` and going "well but this is an integer" even though a sequence of numeric digits is still a valid string.  I need to raise this upstream.  In the meantime we may be able to work around it by massaging the definition of `IntOrString` before we load the cluster Swagger (into a `oneOf` or something like that, I'll have to think about it).

**Comment by itowlson** [comment]
Weirdly I am not seeing this behaviour in development - if I enter an integer for something really stringy like `name` then I get a validation warning as expected, but an integer for `targetPort` doesn't result in a warning.

@brandonkal Could you let me know what OS you are on, and also check the version of the YAML extension that you have installed (via the VS Code Extensions tab), please?  Trying to isolate why you are seeing this and I'm not, so that I can reproduce it and verify a fix.  Thanks!

**Comment by brandonkal** [comment]
MacOS 10.12.6 SSH remote to Ubuntu 18.014

Here are the specs:

VSCode version reported on MacOS:
Version: 1.39.2
Commit: 6ab598523be7a800d7f3eb4d92d7ab9a66069390
Date: 2019-10-15T15:33:00.827Z
Electron: 4.2.10
Chrome: 69.0.3497.128
Node.js: 10.11.0
V8: 6.9.427.31-electron.0
OS: Darwin x64 16.7.0

I don't recall specifically installing this one:
Name: YAML
Id: redhat.vscode-yaml
Description: YAML Language Support by Red Hat, with built-in Kubernetes and Kedge syntax support
Version: 0.5.3
Publisher: Red Hat
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml

Name: Kubernetes
Id: ms-kubernetes-tools.vscode-kubernetes-tools
Description: Develop, deploy and debug Kubernetes applications
Version: 1.0.4
Publisher: Microsoft
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=ms-kubernetes-tools.vscode-kubernetes-tools

**Comment by squillace** [comment]
I have 18.04; lemme give this a shot.

**Comment by squillace** [comment]
Can't repro:
![image](https://user-images.githubusercontent.com/1422718/67051853-51737980-f10a-11e9-8f6c-ac62b407ea25.png)

**Comment by itowlson** [comment]
Might be specific to Mac then - the other report of this for which I have an OS was also Mac.  Or that might be a red herring.

**Comment by squillace** [comment]
or could the kube version affect this at all, on client or server?

**Comment by itowlson** [comment]
@brandonkal Could I also ask you for the version of Kubernetes running in the cluster?  Thanks!

**Comment by brandonkal** [comment]
1.16 on microk8s install

**Comment by brandonkal** [comment]
I have no kubernetes installed on the Mac. It doesn't report an error for local files. So it is either microk8s specific or SSH extension related.

**Comment by itowlson** [comment]
Hmm, okay.  It should be using the version of the active cluster for both local files and live resources, but if the k8s client/connection is on the Ubuntu side and the Mac doesn't have a local kubeconfig then I guess it will be using the fallback schema (which is packaged with the extension) for those and we know the fallback schema _doesn't_ have this issue.

I'll spin up a 1.16 cluster and see if I can reproduce it there.

**Comment by itowlson** [comment]
Aargh.  No repro on Windows with with k8s 1.16 on Kind and referring to a live service resource.  So most likely microk8s or Mac related; I hear you about the SSH factor and I'm not ruling it out, but other users reporting similar issues haven't mentioned having SSH in play, so this feels less likely.  Famous last words.

**Comment by itowlson** [comment]
Got a report from @shreyu86 on #621 - he is on Mac, but using EKS and there is no SSH involved.  So I'm currently thinking Mac related and trying to get hold of a colleague to see if we can reproduce it in that environment.

**Comment by itowlson** [comment]
Comment from @RafaelAybar on #621 reports seeing it on Ubuntu too, so it's not Mac related...  https://github.com/Azure/vscode-kubernetes-tools/issues/621#issuecomment-543563706

**Comment by itowlson** [comment]
Got a repro on Ubuntu!  Thank you @RafaelAybar.

**Comment by itowlson** [comment]
The difference between the fallback schema (which behaves correctly) and the schema given by Kubernetes itself is that in the fallback schema we have:

```
        "io.k8s.apimachinery.pkg.util.intstr.IntOrString": {
            "oneOf": [
                {
                    "type": "string"
                },
                {
                    "type": "integer"
                }
            ]
        },
```

whereas live clusters tell us to use:

```
    "io.k8s.apimachinery.pkg.util.intstr.IntOrString": {
      "description": "IntOrString is a type that can hold an int32 or a string.  When used in JSON or YAML marshalling and unmarshalling, it produces or consumes the inner type.  This allows you to have, for example, a JSON field that can accept a name or number.",
      "format": "int-or-string",
      "type": "string"
    },
```

The latter feels semantically less correct, though arguably Red Hat YAML mishandled it by requiring the quotation marks; I wonder if we should raise an issue against upstream k8s to represent it using the `oneOf` form.

In any case, we can massage it when we retrieve it from the cluster.  And now I have a repro I can _actually confirm the massaging works_.

Thanks everyone for your patience.  Hopefully I'll have a fix for you soon.

**Comment by itowlson** [comment]
Fix on its way (#636) - thanks for your patience!

**Comment by itowlson** [comment]
Closed via #636.
